### PR TITLE
feat(e2e): auto-trigger topup when balance check detects low/critical

### PR DESCRIPTION
**Pull request by jasbanza**
## Summary

Auto-trigger the topup workflow when any balance-check job detects low/critical funds, so E2E accounts are replenished without manual intervention.

## Linear
[MER-170](https://linear.app/osmosis/issue/MER-170/feate2e-auto-trigger-topup-when-balance-check-detects-lowcritical)

## Changes

- Balance-check jobs now dispatch `E2E: Topup Test Accounts` (with `dry_run=false`, `--ref stage`) when outcome is `warn` or `fail`
- Dedup check skips dispatch if a topup is already queued/in-progress
- Concurrency group on topup workflow prevents parallel runs
- Topup Slack summary includes a link to the GitHub Actions run
- Fixed `dry_run` input handling for string coercion edge case

## Files

- **frontend-e2e-tests.yml** / **prod-frontend-e2e-tests.yml** -- trigger step + `actions: write`
- **monitoring-limit-geo-e2e-tests.yml** -- trigger step across all 8 balance-check sections
- **e2e-topup-accounts.yml** -- concurrency group + `dry_run` fix
- **topup-accounts.ts** -- workflow run link in Slack summary

## Test plan

- [ ] Balance-check Slack alerts still fire as before
- [ ] Topup dispatches on warn/fail, skips if one is already pending
- [ ] Concurrency group prevents parallel topup runs

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: .github/workflows/monitoring-limit-geo-e2e-tests.yml [hunk 1]
```diff
@@ -840,6 +999,28 @@ jobs:
         env:
           SLACK_WEBHOOK_URL: ${{ secrets.E2E_SLACK_WEBHOOK_BALANCE_ALERTS }}
           SLACK_WEBHOOK_TYPE: INCOMING_WEBHOOK
+      - name: Trigger automatic topup
+        if: steps.balance-check.outcome == 'failure'
+        continue-on-error: true
+        env:
```
[the remark this answers is not in the document]
Comment (jasbanza) [reply .github/workflows/monitoring-limit-geo-e2e-tests.yml:1040]: implementing a dedup check

File: .github/workflows/monitoring-limit-geo-e2e-tests.yml [hunk 2]
```diff
@@ -78,6 +83,33 @@ jobs:
...
+                  -f reserve_osmo=5 \
+                  -f reserve_usdc=500
+                echo "Topup workflow dispatched (balance outcome: $OUTCOME)."
+              fi
+            else
+              echo "Skipping topup — balance outcome: $OUTCOME"
+            fi
+          else
+            echo "Skipping topup — no balance report found."
+          fi
```
[the remark this answers is not in the document]
Comment (jasbanza) [reply .github/workflows/monitoring-limit-geo-e2e-tests.yml:112]: Acknowledged — the gh run list check is best-effort dedup, not a hard guarantee. The concurrency group (e2e-topup, cancel-in-progress: false) is the actual safety net: at most one pending run is kept, so even if multiple jobs dispatch simultaneously, only 2 runs execute (one active + one queued). The topup script is also idempotent (re-fetches balances, skips funded accounts), so a duplicate run is a no-op. Acceptable trade-off for the simplicity.

File: .github/workflows/frontend-e2e-tests.yml [hunk 3]
```diff
@@ -92,6 +95,33 @@ jobs:
...
+              PENDING=$(gh run list --workflow "E2E: Topup Test Accounts" --status queued --status in_progress --json databaseId --jq 'length')
+              if [ "$PENDING" -gt 0 ]; then
+                echo "Topup workflow already queued/in-progress — skipping dispatch."
+              else
+                gh workflow run "E2E: Topup Test Accounts" \
+                  --ref stage \
+                  -f dry_run=false \
+                  -f topup_multiplier=1.5 \
+                  -f reserve_osmo=5 \
+                  -f reserve_usdc=500
```
[the remark this answers is not in the document]
Comment (jasbanza) [reply .github/workflows/frontend-e2e-tests.yml:116]: Good catch — fixed in 789bb87. Changed inputs.dry_run && 'true' || 'false' to inputs.dry_run == true || inputs.dry_run == 'true' so it handles both boolean (UI dispatch) and potential string coercion (CLI/API dispatch) safely. Also fixed the same pattern in the job name expression.

**Comment by jasbanza** [comment]
@JohnnyWyles Claude opus addressing above review comments by copilot

---

There are **21 Copilot inline comments** across 4 review rounds. They boil down to **5 distinct concerns**, most of which have already been addressed. Here's the breakdown:

### 1. Guard dispatch against transient failures (Comments 1, 3, 5, 7, 8, 10-14)
**Concern:** Topup dispatches on any non-zero exit from `check-balances`, including script errors where no `balance-report.json` is produced.

**Status: Already fixed.** Every dispatch block now has `if [ -f packages/e2e/balance-report.json ]` and checks `OUTCOME` is `warn` or `fail` before dispatching. This was done in commit `e875eea9f`.

### 2. Add `actions: write` permissions (Comments 2, 4, 6, 16, 20)
**Concern:** `gh workflow run` needs `actions: write` on the token.

**Status: Already fixed.**
- `frontend-e2e-tests.yml`: job-level `permissions: { actions: write, contents: read }` on `check-balances`
- `prod-frontend-e2e-tests.yml`: same job-level permissions
- `monitoring-limit-geo-e2e-tests.yml`: workflow-level `permissions: { actions: write, contents: read, deployments: write }`

Comments 16 and 20 say the monitoring workflow should use **job-level** permissions instead of workflow-level to follow least-privilege. **This is a valid nitpick but low risk** -- the workflow is a scheduled internal cron job (not triggered by external PRs), and all 8 jobs in it need the dispatch step, so moving to job-level would mean adding `permissions` to every single job. The `deployments: write` is needed by the `delete-deployments` job. This is a trade-off of verbosity vs. marginal security improvement.

### 3. Dedup dispatches / race condition (Comments 15, 18)
**Concern:** Multiple parallel jobs can all dispatch topup, causing redundant runs.

**Status: Already fixed (best-effort).** Every dispatch block now does a `gh run list --workflow "E2E: Topup Test Accounts" --status queued --status in_progress` check before dispatching, plus the topup workflow has `concurrency: { group: e2e-topup, cancel-in-progress: false }`. Comment 18 acknowledges the TOCTOU race is inherent, and the concurrency group is the real safeguard. This is as good as it gets without a major restructuring.

### 4. Extract duplicated dispatch block (Comments 9, 17, 21)
**Concern:** The "Trigger automatic topup" bash block is duplicated across many jobs/workflows.

**Status: Not addressed, but intentionally so.** Extracting to a reusable workflow or composite action is a reasonable suggestion for long-term maintainability but is a non-trivial refactor (reusable workflows have limitations with secrets, composite actions need a separate directory, etc.). The dispatch block is stable and unlikely to change frequently. **This is a "nice to have" improvement, not a bug.**

### 5. Security: dispatch on untrusted branches (Comment 19)
**Concern:** `frontend-e2e-tests.yml` runs on pushes to non-master branches. A branch could modify the balance check to force repeated topups.

**Status: Not addressed, but low risk in practice.** The topup workflow:
- Uses `--ref stage` (always dispatches from the trusted `stage` branch, not the pushed branch)
- The topup script itself runs from `stage`, not the attacker's branch
- The `balance-report.json` is produced by `check-balances` from the pushed branch, but the worst case is an unnecessary (idempotent) topup
- Only repo members can push branches
- The `continue-on-error: true` + concurrency group limits blast radius

**This is a valid theoretical concern but not a practical risk for this repo.**

---

**Summary: No action needed.** The substantive issues (guard, permissions, dedup, dry_run) were all addressed in earlier commits. The remaining comments are either nitpicks (job-level vs. workflow-level permissions), architectural suggestions (extract reusable action), or theoretical security concerns (untrusted branch dispatch) that don't warrant changes given the repo's threat model.

File: .github/workflows/monitoring-limit-geo-e2e-tests.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasbanza) [reply .github/workflows/monitoring-limit-geo-e2e-tests.yml:112]: see https://github.com/osmosis-labs/osmosis-frontend/pull/4306#issuecomment-4091879832

File: .github/workflows/monitoring-limit-geo-e2e-tests.yml [hunk 4]
```diff
@@ -5,6 +5,11 @@ on:
   schedule:
     - cron: "1 * * * *"
 
+permissions:
+  actions: write
+  contents: read
+  deployments: write
+
```
[the remark this answers is not in the document]
Comment (jasbanza) [reply .github/workflows/monitoring-limit-geo-e2e-tests.yml:12]: see https://github.com/osmosis-labs/osmosis-frontend/pull/4306#issuecomment-4091879832

File: .github/workflows/frontend-e2e-tests.yml [hunk 5]
```diff
@@ -92,6 +95,33 @@ jobs:
...
+                echo "Topup workflow dispatched (balance outcome: $OUTCOME)."
```
[the remark this answers is not in the document]
Comment (jasbanza) [reply .github/workflows/frontend-e2e-tests.yml:117]: see https://github.com/osmosis-labs/osmosis-frontend/pull/4306#issuecomment-4091879832

**Review by JohnnyWyles — APPROVED** [review]
